currency/forexprovider/forexprovider.go

```diff
@@ -24,47 +25,86 @@ func GetAvailableForexProviders() []string {
 
 // NewDefaultFXProvider returns the default forex provider (currencyconverterAPI)
 func NewDefaultFXProvider() *ForexProviders {
-	fxp := new(ForexProviders)
-	currencyC := new(exchangerates.ExchangeRates)
-	currencyC.PrimaryProvider = true
-	currencyC.Enabled = true
-	currencyC.Name = "ExchangeRates"
-	fxp.IFXProviders = append(fxp.IFXProviders, currencyC)
-	return fxp
+	handler := new(ForexProviders)
+	provider := new(exchangerates.ExchangeRates)
+	provider.Setup(base.Settings{
+		PrimaryProvider: true,
+		Enabled:         true,
+		Name:            "ExchangeRates",
+	})
+
+	currencies, _ := provider.GetSupportedCurrencies()
+	providerBase := base.Provider{
+		Provider:            provider,
+		SupportedCurrencies: currencies,
+	}
+
+	handler.FXHandler = base.FXHandler{
+		Primary: providerBase,
+	}
+
+	return handler
+}
+
+// SetProvider sets provider to the FX handler
+func (f *ForexProviders) SetProvider(b base.IFXProvider) error {
+	currencies, err := b.GetSupportedCurrencies()
+	if err != nil {
+		return err
+	}
+
+	providerBase := base.Provider{
+		Provider:            b,
+		SupportedCurrencies: currencies,
+	}
+
+	if b.IsPrimaryProvider() {
+		f.FXHandler = base.FXHandler{
+			Primary: providerBase,
+		}
+		return nil
+	}
+
+	f.FXHandler.Support = append(f.FXHandler.Support, providerBase)
+	return nil
 }
 
 // StartFXService starts the forex provider service and returns a pointer to it
-func StartFXService(fxProviders []base.Settings) *ForexProviders {
-	fxp := new(ForexProviders)
+func StartFXService(fxProviders []base.Settings) (*ForexProviders, error) {
+	handler := new(ForexProviders)
+
 	for i := range fxProviders {
-		if fxProviders[i].Name == "CurrencyConverter" && fxProviders[i].Enabled {
-			currencyC := new(currencyconverter.CurrencyConverter)
-			currencyC.Setup(fxProviders[i])
-			fxp.IFXProviders = append(fxp.IFXProviders, currencyC)
-		}
-		if fxProviders[i].Name == "CurrencyLayer" && fxProviders[i].Enabled {
-			currencyLayerP := new(currencylayer.CurrencyLayer)
-			currencyLayerP.Setup(fxProviders[i])
-			fxp.IFXProviders = append(fxp.IFXProviders, currencyLayerP)
-		}
-		if fxProviders[i].Name == "ExchangeRates" && fxProviders[i].Enabled {
-			exchangeRatesP := new(exchangerates.ExchangeRates)
-			exchangeRatesP.Setup(fxProviders[i])
-			fxp.IFXProviders = append(fxp.IFXProviders, exchangeRatesP)
-		}
-		if fxProviders[i].Name == "Fixer" && fxProviders[i].Enabled {
-			fixerP := new(fixer.Fixer)
-			fixerP.Setup(fxProviders[i])
-			fxp.IFXProviders = append(fxp.IFXProviders, fixerP)
-		}
-		if fxProviders[i].Name == "OpenExchangeRates" && fxProviders[i].Enabled {
-			OpenExchangeRatesP := new(openexchangerates.OXR)
-			OpenExchangeRatesP.Setup(fxProviders[i])
-			fxp.IFXProviders = append(fxp.IFXProviders, OpenExchangeRatesP)
+		switch {
+		case fxProviders[i].Name == "CurrencyConverter" && fxProviders[i].Enabled:
```

Comment:
this feels a duplication of what is done in func (s *Storage) RunUpdater() (currency_storage.go)

Reply:
In currency_storage.go I have these as overridable configs so we can run them without changing configuration enabled = true from false.